katsdptelstate/rdb_writer.py

```diff
@@ -16,145 +16,146 @@
             f.write(RDB_TERMINATOR + RDB_CHECKSUM)
-        if not keys_written:
-            self.logger.error("No valid keys found - removing empty file")
-            os.remove(filename)
-        return (keys_written, keys_failed)
 
-    def encode_supplemental_keys(self, client, keys):
-        """Helper class to encode a specific set of keys on a non-prime Redis like client.
-        """
-        return [self._encode_item(key, client.dump(key)) for key in keys]
 
-    def encode_item(self, key):
-        return self._encode_item(key, self._r.dump(key))
+class RDBWriter(object):
+    """RDB file resource that stores keys from one (or more) Redis DBs.
 
-    def _encode_item(self, key, key_dump):
-        """Returns a binary string containing an RDB encoded key and value.
+    Parameters
+    ----------
+    filename : str
+        Destination filename. Will be opened in 'wb'.
 
-        First byte is used to indicate the encoding used for the value of this key.
-        This is essentially just the Redis type.
+    Attributes
+    ----------
+    keys_written : int
+        Number of keys written to the file.
+    keys_failed : int
+        Number of keys that failed to be written.
+    """
+    def __init__(self, filename):
+        self.filename = filename
+        self.keys_written = 0
+        self.keys_failed = 0
+        self._rdb_context = _rdb_file(self.filename)
+        self._fileobj = None
+
+    def __enter__(self):
+        """Enter RDB writer context, opening RDB file and writing header."""
+        self._fileobj = self._rdb_context.__enter__()
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """Exit RDB writer context, closing off file and deleting it if empty."""
+        try:
+            self._rdb_context.__exit__(exc_type, exc_value, traceback)
+        finally:
+            if not self.keys_written:
+                logger.error("No valid keys found - removing empty file")
+                os.remove(self.filename)
+        return False
 
-        Subsequent bytes represent the name of the key, which consists of a length
-        encoding followed by the actual byte representation of the string. For this simple
-        writer we only provide two length encoding schemas:
-             String length up to 63 - single byte, 6 LSB encode number directly
-             String length from 64 to 16383 - two bytes, 6 LSB of byte 1 + 8 from byte 2 encode number
+    def save(self, client, keys=None):
+        """Save a specified subset of keys from a Redis DB to the RDB dump file.
 
-        Thereafter the value, encoding according to its appropriate schema is appended. As a shortcut,
-        the Redis DUMP command is used to generate the encoded value string.
+        This is a very limited RDB dump utility. Either a compatible redis client
+        or a :class:`katsdptelstate.Backend` must be provided. Missing or broken
+        keys are skipped and logged without raising an exception.
```

Comment:
I think requiring a backend was a historical artefact of the way the code evolved. We could make the interface accept a `TelescopeState` and retrieve the backend from it. It's probably still a reasonable idea for the implementation to work directly on the backend so that everything stays encoded (both the character set sense and the msgpack sense) throughout.

Reply:
Good idea, done.